geode-core/src/test/java/org/apache/geode/cache/query/internal/MethodDispatchTest.java

```diff
@@ -18,64 +18,244 @@
+    queryExecutionContext = new QueryExecutionContext(null, mock(InternalCache.class));
   }
 
   @Test
   public void invokeShouldReturnCorrectlyWhenMethodIsAuthorized()
       throws NameResolutionException, QueryInvocationTargetException {
     doReturn(true).when(methodInvocationAuthorizer).authorize(any(), any());
     MethodDispatch methodDispatch = new MethodDispatch(methodInvocationAuthorizer, TestBean.class,
-        "publicMethod", Collections.emptyList());
+        PUBLIC_METHOD_NAME, Collections.emptyList());
 
-    Object result = methodDispatch.invoke(testBean, emptyList);
+    Object result = methodDispatch.invoke(testBean, emptyList, queryExecutionContext);
     assertThat(result).isInstanceOf(String.class);
-    assertThat(result).isEqualTo("publicMethod");
+    assertThat(result).isEqualTo(PUBLIC_METHOD_RETURN_VALUE);
   }
 
   @Test
   public void invokeShouldThrowExceptionWhenMethodIsNotAuthorized() throws NameResolutionException {
     doReturn(false).when(methodInvocationAuthorizer).authorize(any(), any());
     MethodDispatch methodDispatch = new MethodDispatch(methodInvocationAuthorizer, TestBean.class,
-        "publicMethod", Collections.emptyList());
+        PUBLIC_METHOD_NAME, Collections.emptyList());
 
-    assertThatThrownBy(() -> methodDispatch.invoke(testBean, emptyList))
+    assertThatThrownBy(() -> methodDispatch.invoke(testBean, emptyList, queryExecutionContext))
         .isInstanceOf(NotAuthorizedException.class)
-        .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "publicMethod");
+        .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + PUBLIC_METHOD_NAME);
+  }
+
+  @Test
+  public void invokeShouldCacheResultForAllowedMethod() throws Exception {
+    doReturn(true).when(methodInvocationAuthorizer).authorize(any(), any());
+    MethodDispatch methodDispatch =
+        new MethodDispatch(methodInvocationAuthorizer, Integer.class, "toString",
+            Collections.emptyList());
+    methodDispatch.invoke(new Integer("0"), emptyList, queryExecutionContext);
+
+    assertThat(queryExecutionContext.cacheGet("java.lang.Integer.toString")).isEqualTo(true);
+  }
+
+  @Test
+  public void invokeShouldCacheResultForForbiddenMethod() throws Exception {
+    doReturn(false).when(methodInvocationAuthorizer).authorize(any(), any());
+    MethodDispatch methodDispatch =
+        new MethodDispatch(methodInvocationAuthorizer, Integer.class, "toString",
+            Collections.emptyList());
+
+    assertThatThrownBy(
+        () -> methodDispatch.invoke(new Integer("0"), emptyList, queryExecutionContext))
+            .isInstanceOf(NotAuthorizedException.class)
+            .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "toString");
+
+    assertThat(queryExecutionContext.cacheGet("java.lang.Integer.toString")).isEqualTo(false);
+  }
+
+  @Test
+  public void invokeShouldUseCachedAuthorizerResultWhenMethodIsAuthorizedAndQueryContextIsTheSame()
```

Comment:
random thought... is it possible for someone to access a cq and grab it's query context and somehow used the cached values in it's query context to execute methods/field it shouldn't have access to?